code/lib/core-events/src/errors/server-errors.ts

```diff
@@ -394,6 +395,99 @@ export class NoMatchingExportError extends StorybookError {
   }
 }
 
+export class MainFileESMOnlyImportError extends StorybookError {
+  readonly category = Category.CORE_SERVER;
+
+  readonly code = 5;
+
+  public documentation =
+    'https://github.com/storybookjs/storybook/issues/23972#issuecomment-1948534058';
+
+  constructor(
+    public data: { location: string; line: string | undefined; num: number | undefined }
+  ) {
+    super();
+  }
+
+  template() {
+    const message = [
+      `Storybook failed to load ${this.data.location}..`,
+      '',
+      `It looks like the file tried to load/import an ESM only module.`,
+      `Support for this is currently limited in ${this.data.location}.`,
+      `You can import ESM modules in your main file, but only as dynamic import.`,
+      '',
+    ];
+    if (this.data.line) {
+      message.push(
+        white(
+          `In your ${yellow(this.data.location)} file, line ${bold.cyan(
+            this.data.num
+          )} threw an error, which looks like this:`
+        ),
+        grey(this.data.line)
+      );
+    }
+
+    message.push(
+      '',
+      white(`Convert the static import to an dynamic import ${underline('where they are used')}.`),
+      white(`Example:`) + ' ' + gray(`await import(<your ESM only module>);`),
+      '',
+      'For more information, please read the documentation link below.'
```

Comment:
You don't need to mention it because the error message will already contain a `More info: <link>`
```suggestion
```